resource/readers/resource_namespace_remapper.cpp

```diff
@@ -0,0 +1,212 @@
+ *                   Private Resource Namespace Remapper API                    *
+ *                                                                              *
+ ********************************************************************************/
+
+resource_namespace_remapper_t
+    ::distinct_range_t::distinct_range_t (uint64_t point)
+{
+    m_low = point;
+    m_high = point;
+}
+
+resource_namespace_remapper_t
+    ::distinct_range_t::distinct_range_t (uint64_t lo, uint64_t hi)
+{
+    m_low = lo;
+    m_high = hi;
+}
+
+uint64_t resource_namespace_remapper_t
+             ::distinct_range_t::get_low () const
+{
+    return m_low;
+}
+
+uint64_t resource_namespace_remapper_t
+             ::distinct_range_t::get_high () const
+{
+    return m_high;
+}
+
+bool resource_namespace_remapper_t
+         ::distinct_range_t::is_point () const
+{
+    return m_low == m_high;
+}
+
+bool resource_namespace_remapper_t
+         ::distinct_range_t::operator< (const distinct_range_t &o) const {
+    // this class is used as key for std::map, which relies on the following
+    // x and y are equivalent if !(x < y) && !(y < x)
+    return m_high < o.m_low;
+}
+
+bool resource_namespace_remapper_t
+         ::distinct_range_t::operator== (const distinct_range_t &o) const {
+    // x and y are equal if and only if both high and low
+    // of the operands are equal
+    return m_high == o.m_high && m_low == o.m_low;
+}
+
+bool resource_namespace_remapper_t
+         ::distinct_range_t::operator!= (const distinct_range_t &o) const {
+    // x and y are not equal if either high or low is not equal
+    return m_high != o.m_high || m_low != o.m_low;
+}
+
+
+/********************************************************************************
+ *                                                                              *
+ *                    Public Resource Namespace Remapper API                    *
+ *                                                                              *
+ ********************************************************************************/
+
+int resource_namespace_remapper_t::add (const uint64_t low, const uint64_t high,
+                                        const std::string &name_type,
+                                        uint64_t ref_id, uint64_t remapped_id)
+{
+    if (low > high)
+        goto inval;
+    try {
+        const distinct_range_t exec_target_range{low, high};
+        auto m_remap_iter = m_remap.find (exec_target_range);
+
+        if (m_remap_iter == m_remap.end ()) {
+            auto ret = m_remap.emplace (exec_target_range,
+                                        std::map<const std::string,
+                                                 std::map<uint64_t,
+                                                          uint64_t>> ());
+            if (!ret.second) {
+                errno = ENOMEM;
+                goto error;
+            }
```

Comment:
Since you are using emplace, it is my understanding that you don't need the `find` or the `if`:

```suggestion
        auto ret = m_remap.emplace (exec_target_range,
                                    std::map<const std::string,
                                                std::map<uint64_t,
                                                         uint64_t>> ());
```

Also IIUC, `insert` and `emplace` only return false for `ret.second` when the element already existed in the map, not when there is a memory error, so if you keep those checks, I think `ENOMEM` is a misnomer.  `EEXISTS` might be better.  I'm basing this off the docs in cplusplus.com, so I very well may be wrong about that. [[1]](https://www.cplusplus.com/reference/map/map/emplace/) [[2]](https://www.cplusplus.com/reference/map/map/insert/)

Reply:
Sorry, I see now that you are using the result of the find further down. So ignore my comment about not needing the `find`. Sorry about that.

The bit about `ENOMEM` still stands though.  Assuming that `ret.second` is only false when the element already exists (and not when out of memory), I don't think you need that check at all (the find already proved the element doesn't exist).

Reply:
I considered `EEXIST` but that condition will never occur because of the enclosing condition: `if (m_remap_iter == m_remap.end ()) {`

So I used `ENOMEM` as catch-all. Maybe it is better to just not checking `ret.second` since it should not occur?

Reply:
> Maybe it is better to just not checking ret.second since it should not occur?

Yeah, I would agree with that.  Your call though.  My previous approval still stands, so feel free to put MWP on this PR when you are happy with it.  It'll probably need a rebase too for the new GitHub actions to take affect (I'm still seeing Travis on the status of this).